cmd/update.go

```diff
@@ -0,0 +1,483 @@
+package cmd
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	progress "github.com/oasisprotocol/cli/cmd/common"
+
+	cliVersion "github.com/oasisprotocol/cli/version"
+	"github.com/spf13/cobra"
+)
+
+const (
+	apiTimeout      = 15 * time.Second
+	downloadTimeout = 5 * time.Minute
+	maxChecksumSize = 1024 // bytes
+)
+
+const (
+	osWindows     = "windows"
+	binaryName    = "oasis"
+	binaryNameWin = binaryName + ".exe"
+)
+
+var httpClient = &http.Client{
+	Transport: &http.Transport{
+		TLSHandshakeTimeout:   10 * time.Second,
+		ResponseHeaderTimeout: 30 * time.Second,
+		Proxy:                 http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   10 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+	},
+}
+
+var (
+	githubAPIBase = "https://api.github.com"
+
+	downloadFunc = doDownload
+
+	osExecutable = os.Executable
+)
+
+// semverDiffers reports whether versions a and b differ (ignoring leading 'v').
+func semverDiffers(a, b string) (bool, error) {
+	normalize := func(s string) string {
+		s = strings.TrimPrefix(s, "v")
+		// Strip pre-release (-…) and build metadata (+…)
+		if idx := strings.IndexAny(s, "+-"); idx != -1 {
+			s = s[:idx]
+		}
+		return s
+	}
+
+	aNorm := normalize(a)
+	bNorm := normalize(b)
+
+	if aNorm == "" || bNorm == "" {
+		return false, fmt.Errorf("invalid version string (empty)")
+	}
+	return aNorm != bNorm, nil
+}
+
+type ghRelease struct {
```

Comment:
In the comment add a link to github's API reference that describes this structure.